File: ts_webgl/src/demo/SolidColorState.ts
import { Rgba, Size2 } from "../geometry";
import { AppState } from "../state-machine";

export class SolidColorState implements AppState<WebGL2RenderingContext> {
	private size = new Size2(0, 0);

	constructor(readonly color: Rgba) { }

	activate(_context: AppState.Context<WebGL2RenderingContext>): void {
		// nothing to do
	}

	deactivate(_context: AppState.Context<WebGL2RenderingContext>): void {
		// nothing to do	
	}

	resize(_context: AppState.Context<WebGL2RenderingContext>): void {
		// nothing to do
	}

	render(context: AppState.Context<WebGL2RenderingContext>): void {
		const gl = context.renderingContext;
		gl.viewport(0, 0, this.size.width, this.size.height);
		gl.clearColor(this.color.red, this.color.green, this.color.blue, this.color.alpha);
		gl.clear(gl.COLOR_BUFFER_BIT);
	}

	update(_context: AppState.Context<WebGL2RenderingContext>, _elapsedTime: number): AppState<WebGL2RenderingContext> | null | undefined {
		return null;
	}
}

File: rust_webgl/client/src/solid_color_state.rs
use std::{rc::Rc, time::Duration};

use vek::Extent2;
use vek::Rgba;
use web_sys::WebGl2RenderingContext;

use lib::*;

pub struct SolidColorState {
	color: Rgba<f32>,
	gl: Option<Rc<WebGl2RenderingContext>>,
}

impl SolidColorState {
	pub fn new(color: Rgba<f32>) -> Self {
		Self { color, gl: None }
	}
}

impl AppState for SolidColorState {
	fn activate(&mut self, gl: Rc<WebGl2RenderingContext>) -> AppResult<()> {
		self.gl = Some(gl.clone());
		Ok(())
	}

	fn deactivate(&mut self) -> AppResult<()> {
		Ok(())
	}

	fn resize(&mut self, size: Extent2<i32>) -> AppResult<()> {
		let gl = self.gl.clone().unwrap();

		gl.viewport(0, 0, size.w, size.h);

		Ok(())
	}

	fn render(&mut self) -> AppResult<()> {
		let gl = self.gl.clone().unwrap();

		gl.clear_color(self.color.r, self.color.g, self.color.b, self.color.a);
		gl.clear(WebGl2RenderingContext::COLOR_BUFFER_BIT);

		Ok(())
	}

	fn update(&mut self, _time: Duration) -> AppResult<Option<AppStateHandle>> {
		Ok(None)
	}
}
